New feature: Add schema metadata to deserialized avro records

sbrauer opened the pull request:
~~When an Avro record is deserialized to a Clojure map, add the record's name and fully-qualified name to the map as metadata.
These record names are potentially useful info that was being discarded.~~

When an Avro record is deserialized to a Clojure map, add the record's schema to the map as metadata.

My personal use case for this metadata is to make it easier for consuming code to deal with unions of records.

sbrauer committed:
Add record name metadata to deserialized avro records

When an Avro record is deserialized to a Clojure map, add the record's
name and fully-qualified name to the map as metadata.
These record names are potentially useful info that was being discarded.

My personal use case for this metadata is to make it easier for
consuming code to deal with unions of records.

Signed-off-by: Sam Brauer <sam.brauer@gmail.com>

cddr commented on src/jackdaw/serdes/avro.clj:
```diff
@@ -399,21 +399,24 @@
+                             [field-key (avro->clj field-coercion value)]))))
+              (vals field->schema+coercion))
+        {:name (.getName schema)
+         :fullname (.getFullName schema)})))
```
I think when hyphenating the `FullName` symbol I'd expect to see `full-name`.

sbrauer replied:
I just pushed a commit with the hyphenated name.

sbrauer committed:
Rename :fullname to :full-name

Signed-off-by: Sam Brauer <sam.brauer@gmail.com>

sbrauer commented:
I hope you agree that there's value in exposing the record names. There may be other ways to do that besides metadata (although it was the best way I could think of), and even with the metadata approach one might have chosen different keywords.
I'd be happy to discuss.

cddr commented:
Hey Sam. Sorry for the radio silence so far. The other person in our company who might have an opinion on this is away on holiday at the moment.

I agree that this is valuable information that should be made available and that given the current setup, the metadata seems like the only place to put it. I wonder whether we could just put the whole schema in the metadata (or maybe even both a writer schema and a reader schema).

sbrauer committed:
Add schema object to record metadata

Adding the record names only was probably too limiting and opinionated.
By exposing the record schema object a consumer has more flexibility
to get the names and lots of other data besides.

Signed-off-by: Sam Brauer <sam.brauer@gmail.com>

sbrauer commented:
No worries. Good to hear that you're open to the general idea. I've also been thinking that exposing the schema itself could be better than exposing the names (since the names and a lot of other info can be obtained from the schema; on the other hand one could argue it's exposing too much internal detail).

I must admit that I haven't yet taken the time to digest your recent change related to decoupling the reader and writer schemas. I should do that soon. In the meantime I pushed a commit that adds the schema to the metadata instead of the names. I'll update the PR title as well.
